src/core.rs

```diff
@@ -0,0 +1,754 @@
+                }
+
+                let var = fresh_gen.fresh(f);
+                binding.insert(var.clone());
+
+                norm_actions.push(GenericCoreAction::Let(var.clone(), f.clone(), norm_args));
+                Ok((
+                    GenericCoreActions::new(norm_actions),
+                    GenericExpr::Call((), (f.clone(), var), mapped_args),
+                ))
+            }
+        }
+    }
+}
+
+/// A [`GenericCoreRule`] represents a generalization of lowered form of a rule.
+/// Unlike other `Generic`-prefixed types, [`GenericCoreRule`] takes two `Head`
+/// parameters instead of one. This is because the `Head` parameter of `body` and
+/// `head` can be different. In particular, early in the compilation pipeline,
+/// `body` can contain `Eq` atoms, which denotes equality constraints, so the `Head`
+/// for `body` needs to be a `HeadOrEq<Head>`, while `head` does not have equality
+/// constraints.
+#[derive(Debug, Clone)]
+pub struct GenericCoreRule<HeadQ, HeadA, Leaf> {
+    pub body: Query<HeadQ, Leaf>,
+    pub head: GenericCoreActions<HeadA, Leaf>,
+}
+
+pub(crate) type CoreRule = GenericCoreRule<SymbolOrEq, Symbol, Symbol>;
+pub(crate) type ResolvedCoreRule = GenericCoreRule<ResolvedCall, ResolvedCall, ResolvedVar>;
+
+impl<Head1, Head2, Leaf> GenericCoreRule<Head1, Head2, Leaf>
+where
+    Head1: Clone,
+    Head2: Clone,
+    Leaf: Clone + Eq + Hash,
+{
+    pub fn subst(&mut self, subst: &HashMap<Leaf, GenericAtomTerm<Leaf>>) {
+        for atom in &mut self.body.atoms {
+            atom.subst(subst);
+        }
+        self.head.subst(subst);
+    }
+}
+
+impl<Head, Leaf> GenericCoreRule<HeadOrEq<Head>, Head, Leaf>
+where
+    Leaf: Eq + Clone + Hash + Debug,
+    Head: Clone,
+{
+    /// Transformed a UnresolvedCoreRule into a CanonicalizedCoreRule.
+    /// In particular, it removes equality checks between variables and
+    /// other arguments, and turns equality checks between non-variable arguments
+    /// into a primitive equality check `value-eq`.
+    pub(crate) fn canonicalize(
+        self,
+        // Users need to pass in a substitute for equality constraints.
+        value_eq: impl Fn(&GenericAtomTerm<Leaf>, &GenericAtomTerm<Leaf>) -> Head,
+    ) -> GenericCoreRule<Head, Head, Leaf> {
+        let mut result_rule = self;
+        loop {
+            let mut to_subst = None;
+            for atom in result_rule.body.atoms.iter() {
+                if atom.head.is_eq() && atom.args[0] != atom.args[1] {
+                    match &atom.args[..] {
+                        [GenericAtomTerm::Var(x), y] | [y, GenericAtomTerm::Var(x)] => {
+                            to_subst = Some((x, y));
+                            break;
+                        }
+                        _ => (),
+                    }
+                }
+            }
+            if let Some((x, y)) = to_subst {
+                let subst = HashMap::from_iter(vec![(x.clone(), y.clone())]);
```

Comment:
nit: I don't think you need `vec!` here (you should be able to iterate over the array, or at least pass a `iter::once(`)